telemetry/specs/INT.mdk

```diff
@@ -1220,6 +1251,128 @@ Metadata if added by the source. The total stack length can be determined
 by subtracting the total INT fixed header sizes (12 bytes)
 from (shim header length \* 4).
 
+## INT-MX Metadata Header Format
+
+In this section, we define the format for INT-MX metadata headers,
+and the metadata itself.
+
+INT-MX Metadata Header:
+`
+    0                   1                   2                   3
+    0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |Ver = 2|D|E|M|                  Reserved                       |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |      Instruction Bitmap       |        Domain Specific ID     |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   |         DS Instruction        |            DS Flags           |
+   +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+`
+* INT metadata header is 12 bytes long. Each metadata requested in the 
+INT Header instruction is either 4 bytes or 8 bytes in length. Each INT 
+capable device in the forwarding path will send the requested metadata
+monitoring system in the Telemetry Report.
+
+Details of the metadata semantics and format of the Telemetry Report can 
+be accessed at the link below:
+  https://github.com/p4lang/p4-applications/blob/master/docs/telemetry_report_v2.0.pdf
+
+* The fields in the INT metadata header are interpreted the following way:
+  - Ver (4b): INT metadata header version. Should be 2 for this version.
+  - D   (1b): Discard Copy/Clone. Not applicable in INT-MX mode. Should be set to 0 
+         by the INT Source.
+  - E   (1b): Max Hop Count exceeded. Not applicable in INT-MX mode. Should be set to 0 
+         by the INT Source.
+  - M   (1b): MTU exceeded. Not applicable in INT-MX mode. Should be set to 0 
+         by the INT Source.
+  - R  (25b): Reserved bits. Should be set to 0 by INT Source and ignored by other devices.
```

Comment:
What if someone implements only INT-MX. This would mean that Source Node has to compute the Metadata requested and program the Hop_ML field. For INT-MX, the node generating the Telemetry Report has to gather the metadata requested and compute the length anyways. So setting the MD Length field in the Telemetry Report shouldn't be an issue. Also, what if the metadata requested is not available at a node, then that node will have to compute the M<D Length accordingly anyways. Hop_ML is critical in INT-MD, not in INT-MX. Thoughts?

Reply:
I am proposing the addition of one requirement at the Source Node, to set MD_Length, and then allowing a variety of methods at transit and sink nodes. If you don't want to look at MD_Length in the INT-MX header at transit and sink, you don't have to.

I am proposing two different implementation options at transit and sink nodes implementing INT-MX:

1. As you mention, when requested metadata is not available or one or more bits are not understood, bits could be cleared when copying into the Telemetry Report's RepMdBits and DSMdBits. In that case MD_Length in the Telemetry Report would need to be recomputed. MD_Length was put into the Telemetry Report to ease in parsing at the collector, in case the collector does not know or understand some bit definitions in DSMdBits, so that the collector can find the start of the inner report contents such as truncated packet fragments. This is purposely meant to be similar to the way Hop_ML in INT-MD is used by the collector to distinguish metadata from different hops.

2. An alternative would be to do what INT-MD does, which is to leave all the bits set, and insert the special all-ones value for unavailable or unsupported metadata. This is why MD_Length must be inserted by the source node, so that these implementations can determine how much padding to add. With this approach, there is no need to recompute MD_Length.

I am not sure I see a reason to preclude a hybrid, where some bits are cleared and others have the special all-ones value, since that would still be parseable at the collector. However, I am not sure why one would do that.